File: AIペアプログラミング実践編_ソースコード/practical-python/session-4/Ans/inventory/services/product_service.py
"""商品サービス"""
from inventory.models.product import Product

from .dto import ProductInput
from .product_validator import ProductValidator


class ProductService:
    """商品のビジネスロジック"""

    def __init__(self, repository=None):
        self.repository = repository
        self.validator = ProductValidator()

    def create(self, input_data: ProductInput) -> Product:
        """商品を登録"""
        # バリデーション
        result = self.validator.validate(input_data)
        if not result.is_valid:
            raise ValueError(", ".join(result.errors))

        # 商品を作成
        product = Product(
            name=input_data.name,
            category=input_data.category,
            price=input_data.price
        )

        # 保存
        if self.repository:
            return self.repository.save(product)
        else:
            product.save()
            return product

    def find_all(self):
        """全商品を取得"""
        if self.repository:
            return self.repository.find_all()
        return list(Product.objects.all())

    def find_by_id(self, product_id: int) -> Product:
        """ID で商品を取得"""
        if self.repository:
            product = self.repository.find_by_id(product_id)
        else:
            product = Product.objects.filter(id=product_id).first()

        if product is None:
            raise ValueError(f"商品が見つかりません: {product_id}")
        return product

    def search(
        self,
        name: str | None = None,
        category: str | None = None,
        min_price: int | None = None,
        max_price: int | None = None
    ):
        """商品を検索"""
        products = self.find_all()

        return [
            product for product in products
            if self._matches_criteria(product, name, category, min_price, max_price)
        ]

    def _matches_criteria(
        self,
        product,
        name: str | None,
        category: str | None,
        min_price: int | None,
        max_price: int | None
    ) -> bool:
        """検索条件にマッチするか判定"""
        # 商品名（部分一致）
        if name and name not in product.name:
            return False

        # カテゴリ（完全一致）
        if category and product.category != category:
            return False

        # 価格帯
        if min_price is not None and product.price < min_price:
            return False
        if max_price is not None and product.price > max_price:
            return False

        return True

File: AIペアプログラミング実践編_ソースコード/practical-java/session-4/Ans/src/main/java/com/example/inventory/service/ProductService.java
package com.example.inventory.service;

import com.example.inventory.dto.ProductInput;
import com.example.inventory.dto.ValidationResult;
import com.example.inventory.entity.Product;
import com.example.inventory.entity.ProductCategory;
import com.example.inventory.repository.ProductRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

/**
 * 商品サービス
 */
@Service
public class ProductService {
    private final ProductRepository productRepository;
    private final ProductValidator productValidator;

    @Autowired
    public ProductService(ProductRepository productRepository) {
        this.productRepository = productRepository;
        this.productValidator = new ProductValidator();
    }

    // テスト用コンストラクタ
    public ProductService(ProductRepository productRepository, ProductValidator productValidator) {
        this.productRepository = productRepository;
        this.productValidator = productValidator;
    }

    public Product create(ProductInput input) {
        // バリデーション
        ValidationResult validation = productValidator.validate(input);
        if (!validation.isValid()) {
            throw new IllegalArgumentException(String.join(", ", validation.getErrors()));
        }

        // エンティティの作成
        Product product = new Product();
        product.setName(input.getName());
        product.setCategory(input.getCategory());
        product.setPrice(input.getPrice());

        return productRepository.save(product);
    }

    public List<Product> findAll() {
        return productRepository.findAll();
    }

    public Product findById(Long id) {
        return productRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("商品が見つかりません: " + id));
    }

    public void delete(Long id) {
        if (!productRepository.existsById(id)) {
            throw new IllegalArgumentException("商品が見つかりません: " + id);
        }
        productRepository.deleteById(id);
    }

    public List<Product> search(String name, ProductCategory category,
                               Integer minPrice, Integer maxPrice) {
        List<Product> products = productRepository.findAll();

        return products.stream()
                .filter(product -> {
                    // 商品名（部分一致）
                    if (name != null && !product.getName().contains(name)) {
                        return false;
                    }
                    // カテゴリ（完全一致）
                    if (category != null && product.getCategory() != category) {
                        return false;
                    }
                    // 価格帯
                    if (minPrice != null && product.getPrice() < minPrice) {
                        return false;
                    }
                    if (maxPrice != null && product.getPrice() > maxPrice) {
                        return false;
                    }
                    return true;
                })
                .toList();
    }
}
